rcrecv-daemon.8
.\"-
.\"Copyright (c) 2021, Alexander Mishin
.\"All rights reserved.
.\"
.\"Redistribution and use in source and binary forms, with or without
.\"modification, are permitted provided that the following conditions are met:
.\"
.\"* Redistributions of source code must retain the above copyright notice, this
.\"  list of conditions and the following disclaimer.
.\"
.\"* Redistributions in binary form must reproduce the above copyright notice,
.\"  this list of conditions and the following disclaimer in the documentation
.\"  and/or other materials provided with the distribution.
.\"
.\"THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\"AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\"IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\"DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
.\"FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\"DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\"SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
.\"CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\"OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\"OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.Dd May 21, 2021
.Dt RCRECV-DAEMON 8
.Os
.Sh NAME
.Nm rcrecv-daemon
.Nd GPIO control daemon for
.Xr rcrecv 4
kernel driver.
.Sh SYNOPSIS
.Nm
.Op Fl b
.Op Fl d Ar device
.Op Fl g Ar device
.Op Fl i Ar ms
.Oo
.Fl s Ns | Ns Fl u Ns | Ns Fl t
.Ar code Ns = Ns Ar value Ns , Ns Ar pin Ns = Ns Ar number ...
.Oc
.Op Fl h
.Sh OPTIONS
The
.Nm
supports the following options:
.Bl -tag -width indent
.It Fl b
Daemonize the
.Nm
after running;
.It Fl d , Ic --device Ar device
Specify a
.Xr rcrecv 4
character device. The compiled-in default is
.Pa /dev/rcrecv Ns ;
.It Fl g , Ic --gpio Ar device
Specify a gpio controller. The compiled-in default is
.Pa /dev/gpioc0 Ns ;
.It Fl i , Ic --interval Ar ms
Specify a minimal interval for ignoring the same codes repeated too fast. Default value is 1000ms;
.It Fl s , Ic --set Ar code Ns = Ns Ar value Ns , Ns Ar pin Ns = Ns Ar number
.It Fl u , Ic --unset Ar code Ns = Ns Ar value Ns , Ns Ar pin Ns = Ns Ar number
.It Fl t , Ic --toggle Ar code Ns = Ns Ar value Ns , Ns Ar pin Ns = Ns Ar number
Set, unset or toggle the gpio
.Ar pin
when the
.Ar code
is received.
.Ar Value
and
.Ar number
must be integer constants, with a leading 0x indicating a hexadecimal value and
a leading 0 indicating an octal value. Options
.Fl s , Fl u
and
.Fl t
can be used multiple times in any order;
.It Fl h , Ic --help
Brief help.
.Sh DESCRIPTION
A
.Nm
needs to have a
.Xr rcrecv 4
loaded for it to work.
.br
When the daemon is started it interprets
.Ic --set , --unset
and
.Ic --toggle
parameters to make a list of requested states of pins for each code to receive.
Then the daemon waits a
.Xr kevent 2
from
.Xr rcrecv 4
which means that there is a code for it. The daemon gets the code, seek a pin
defined for it, configures the pin for output and changes state of that pin
the way have been defined by a corresponding parameter.
.Sh EXAMPLES
An example of a
.Xr service 8
configuration file looks like this:
.Bd -literal -offset indent
rcrecv_daemon_enable="YES"
rcrecv_daemon_codes="0x952e91 0x952e92 0x952e94"
rcrecv_0x952e91_pin=12
rcrecv_0x952e91_state="t"
rcrecv_0x952e92_pin=11
rcrecv_0x952e92_state="s"
rcrecv_0x952e94_pin=11
rcrecv_0x952e94_state="u"
.Ed

It will make the
.Xr service 8
script to run the daemon with following options:
.Bd -literal -offset indent
$ rcrecv-daemon -b -t 0x952e91:12 -s 0x952e92:11 -u 0x952e94:11
.Ed

Note: If You want to make a
.Xr service 8
configuration file in the
.Pa rc.conf.d/
directory, the file must be named
.Ql rcrecv_daemon
(with an underscore, not a hyphen).
.Sh FILES
/usr/local/etc/rc.d/rcrecv-daemon
.Sh SEE ALSO
.Xr rcrecv 4 ,
.Xr gpioctl 8
.Sh AUTHORS
The
.Nm
was written by
.An Alexander Mishin Aq Mt mishin@mh.net.ru .


rcrecv-daemon.c
/*-
 * SPDX-License-Identifier: BSD-2-Clause-FreeBSD
 *
 * Copyright (C) 2021 Alexander Mishin
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * GPIORCRECV - Remote control receiver module over GPIO.
 *
 * Remote control receiver module can't be discovered automatically, please
 * specify hints as part of loader or kernel configuration:
 *	hint.rcrecv.0.at="gpiobus0"
 *	hint.rcrecv.0.pins=<PIN>
 *
 * Or configure via FDT data.
 */

#include <sys/event.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <sys/ioctl.h>
#include <stdbool.h>
#include <unistd.h>
#include <getopt.h>
#include <signal.h>
#include <syslog.h>

#include <libutil.h>
#include <libgpio.h>

#include <dev/rcrecv/rcrecv.h>

#define DEVICE_GPIOC  "/dev/gpioc0"
#define DEVICE_RCRECV "/dev/rcrecv"

static const char *dev_rcrecv = DEVICE_RCRECV;
static const char *dev_gpio = DEVICE_GPIOC;
static int dev;
static gpio_handle_t gpioc;
static struct pidfh *pfh;
static bool background = false;
static unsigned long interval = 1000; // Minimal interval between codes is 1s

typedef struct rcc_entry {
    uint32_t code;
    gpio_pin_t pin;
    char state;

    SLIST_ENTRY(rcc_entry) switches;
} *rcc_entry_t;

SLIST_HEAD(rcc_list, rcc_entry);

static struct rcc_list search_switch;

static void
usage()
{
    fprintf(stderr, "usage: %s [-d <ctldev>] [-g <gpioc>] [-i <ms>]"
	"-(s|u|t) code=<code>,pin=<pin> [-b] [-h]\n\n",
	getprogname());
    fprintf(stderr,
	"Options:\n"
	"    -d, --device=<ctldev>\n"
	"                        A remote control receiver device name.\n"
	"                        Default: /dev/rcrecv;\n"
	"    -g, --gpio=<gpioc>  A gpio controller device name\n"
	"                        Default: /dev/gpioc0;\n"
	"    -i, --interval=<ms> A minimal valid interval between repeated codes\n"
	"                        If an interval is less than that value the next same ones\n"
	"                        will be ignored. Default value is 1000ms;"
	"    -s, --set code=<code>,pin=<pin>\n"
	"    -u, --unset code=<code>,pin=<pin>\n"
	"    -u, --toggle code=<code>,pin=<pin>\n"
	"                        A way which the <pin> should be changed after the\n"
	"                        <code> is received;\n"
	"    -b,                 Run in background as a daemon;\n"
	"    -h, --help          Print this help.\n"
    );

}

/* Signals handler. Prepare the programm for end */
static void
termination_handler(int signum)
{
    /* Free allocated memory blocks */
    while (!SLIST_EMPTY(&search_switch)) {           /* List Deletion. */
	rcc_entry_t node = SLIST_FIRST(&search_switch);
	SLIST_REMOVE_HEAD(&search_switch, switches);
	free(node);
    }

    /* Close devices */
    close(dev);
    gpio_close(gpioc);

    /* Remove pidfile and exit */
    pidfile_remove(pfh);

    exit(EXIT_SUCCESS);
}

/* Daemonize the program */
static void
daemonize(void)
{
    pid_t otherpid;

    /* Try to create a pidfile */
    pfh = pidfile_open(NULL, 0600, &otherpid);
    if (pfh == NULL)
    {
	if (errno == EEXIST)
	    errx(EXIT_FAILURE, "Daemon already running, pid: %jd.", (intmax_t)otherpid);

	/* If we cannot create pidfile from other reasons, only warn. */
	warn("Cannot open or create pidfile");
	/*
	 * Even though pfh is NULL we can continue, as the other pidfile_*
	 * function can handle such situation by doing nothing except setting
	 * errno to EDOOFUS.
	 */
    }

    /* Try to demonize the process */
    if (daemon(0, 0) == -1)
    {
	pidfile_remove(pfh);
	errx(EXIT_FAILURE, "Cannot daemonize");
    }

    pidfile_write(pfh);
}

/* Adds new code map entry */
static rcc_entry_t
search_rcc_entry(const uint32_t *code)
{
    rcc_entry_t node;

    SLIST_FOREACH(node, &search_switch, switches)
    {
	printf("code: %u, pin: %u, state: %c\n", node->code, node->pin, node->state);
	if (node->code == *code)
	    return (node);
    }

    return (NULL);
}

/* Adds new code map entry */
static rcc_entry_t
add_rcc_entry(rcc_entry_t curnode)
{
    rcc_entry_t newnode;

    newnode = (rcc_entry_t)malloc(sizeof(*newnode));
    if (newnode == NULL)
	err(EXIT_FAILURE, "Unable to malloc symbol_node\n");

    if (curnode == NULL)
	SLIST_INSERT_HEAD(&search_switch, newnode, switches);
    else if (SLIST_NEXT(curnode, switches) == NULL)
	SLIST_INSERT_AFTER(curnode, newnode, switches);
    else
	err(EXIT_FAILURE, "Unable to add a switch entry\n");

    return newnode;
}

/* Get and decode params */
static void
get_param(int argc, char **argv)
{
    int ch, long_index = 0;
    rcc_entry_t node;
    extern char *optarg, *suboptarg;
    char *options, *value, *end;

    char *subopts[] = {
#define CODE		0
		"code",
#define PIN		1
		"pin",
	NULL
    };

    static struct option long_options[] = {
//        {"status", no_argument,       0, 'v' },
	{"device",   required_argument, 0, 'd' },
	{"gpio",     required_argument, 0, 'g' },
	{"interval", required_argument, 0, 'i' },
	{"set",      required_argument, 0, 's' },
	{"unset",    required_argument, 0, 'u' },
	{"toggle",   required_argument, 0, 't' },
	{"help",     required_argument, 0, 'h' },
	{0, 0, 0, 0}
    };

    SLIST_INIT(&search_switch);
    node = SLIST_FIRST(&search_switch);

    while ((ch = getopt_long(argc, argv, "d:s:t:u:bh",long_options,&long_index)) != -1) {
	switch (ch) {
	case 'd':
	    dev_rcrecv = optarg;
	    break;
	case 'g':
	    dev_gpio = optarg;
	    break;
	case 'i':
	    interval = strtoul(optarg, &end, 0);
	    break;
	case 'b':
	    background = true;
	    break;
	case 's':
	    /* FALLTHROUGH */
	case 'u':
	    /* FALLTHROUGH */
	case 't':
	    node = add_rcc_entry(node);

	    /* Set a pin state */
	    node->state = ch;

	    options = optarg;
	    while (*options) {
		switch(getsubopt(&options, subopts, &value)) {
		case CODE:
		    if (!value)
			errx(EXIT_FAILURE, "no code");
		    node->code = strtoul(value, &end, 0);
		    break;
		case PIN:
		    if (!value)
			errx(EXIT_FAILURE, "no pin number");
		    node->pin = strtoul(value, &end, 0);
		    break;
		case -1:
		    if (suboptarg)
			errx(EXIT_FAILURE, "illegal sub option %s", suboptarg);
		    else
			errx(EXIT_FAILURE, "no code neither pin number");
		    break;
		}
	    }
	    break;
	case 'h':
	    /* FALLTHROUGH */
	default:
	    usage();
	    exit(EXIT_SUCCESS);
	}
    }
}

int
main(int argc, char **argv)
{
    rcc_entry_t node;

    struct timespec timeout;
    const size_t waitms = 1000;
    int64_t last_time = 0;
    uint32_t last_code = 0;
    char *action;

    struct kevent event;    /* Event monitored */
    struct kevent tevent;   /* Event triggered */
    int kq, ret;
    struct rcrecv_code rcc;

    get_param(argc, argv);

    /* Set a timeout by 'waitms' value */
    timeout.tv_sec = waitms / 1000;
    timeout.tv_nsec = (waitms % 1000) * 1000 * 1000;

    /* Open GPIO controller */
    gpioc = gpio_open_device(dev_gpio);
    //gpio_pin_output(handle, 16);
    if (gpioc == GPIO_INVALID_HANDLE)
	errx(EXIT_FAILURE, "Failed to open '%s'", dev_gpio);

    /* Open RCRecv device */
    dev = open(dev_rcrecv, O_RDONLY);
    if (dev < 0)
	errx(EXIT_FAILURE, "Failed to open '%s'", dev_rcrecv);

    /* Create kqueue. */
    kq = kqueue();
    if (kq == -1)
	err(EXIT_FAILURE, "kqueue() failed");

    /* Initialize kevent structure. */
    EV_SET(&event, dev, EVFILT_READ, EV_ADD | EV_CLEAR, 0, 0, NULL);
    /* Attach event to the kqueue. */
    ret = kevent(kq, &event, 1, NULL, 0, NULL);
    if (ret == -1)
	err(EXIT_FAILURE, "kevent register");
    if (event.flags & EV_ERROR)
	errx(EXIT_FAILURE, "Event error: %s", strerror(event.data));

    /* If there are dirty kevents read and drop irrelevant data */
    ret = kevent(kq, NULL, 0, &tevent, 1, &timeout);
    if (ret == -1) {
	err(EXIT_FAILURE, "kevent wait");
    }
    else if (ret > 0) {
	ioctl(dev, RCRECV_READ_CODE_INFO, &rcc);
    }

    /* Unbinds from terminal if '-b' */
    if (background) {
	daemonize();

	/* Create kqueue for child */
	kq = kqueue();
	if (kq == -1)
	    err(EXIT_FAILURE, "kqueue() failed");

	/* Initialize kevent structure once more */
	EV_SET(&event, dev, EVFILT_READ, EV_ADD | EV_CLEAR, 0, 0, NULL);
	/* and  once more attach event to the kqueue. */
	ret = kevent(kq, &event, 1, NULL, 0, NULL);
	if (ret == -1)
	    err(EXIT_FAILURE, "kevent register");
	if (event.flags & EV_ERROR)
	    errx(EXIT_FAILURE, "Event error: %s", strerror(event.data));
    }

    /* Intercept signals to our function */
    if (signal (SIGINT, termination_handler) == SIG_IGN)
	signal (SIGINT, SIG_IGN);
    if (signal (SIGTERM, termination_handler) == SIG_IGN)
	signal (SIGTERM, SIG_IGN);

    for (;;) {
	/* Sleep until a code received */
	ret = kevent(kq, NULL, 0, &tevent, 1, &timeout);
	if (ret == -1) {
	    err(EXIT_FAILURE, "kevent wait");
	}
	else if (ret > 0) {
	    /* Get a code from remote control
	       and search a node for it */
	    ioctl(dev, RCRECV_READ_CODE_INFO, &rcc);
	    node = search_rcc_entry(&rcc.value);
	    /* No actions if the same code is repeated too fast */
	    if (node != NULL &&
	       (last_code != rcc.value ||
		llabs(last_time - rcc.last_time) > (interval * 1000)))
	    {
		/* Config a pin from the node for output before change it */
		gpio_pin_output(gpioc, node->pin);
		/* Change state of the pin as set in the node */
		switch(node->state) {
		case 's':
		    action = "Set";
		    gpio_pin_high(gpioc, node->pin);
		    break;
		case 'u':
		    action = "Unset";
		    gpio_pin_low(gpioc, node->pin);
		    break;
		case 't':
		    action = "Toggle";
		    gpio_pin_toggle(gpioc, node->pin);
		    break;
		}
		last_time = rcc.last_time;
		last_code = rcc.value;
		syslog(LOG_INFO, "Received code 0x%X: %s pin %u\n",
			node->code, action, node->pin);
	    }
	}
    }
}
